server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/RangeAggregator.java

```diff
@@ -312,29 +372,195 @@ private int collect(int doc, double value, long owningBucketOrdinal, int lowBoun
 
         for (int i = startLo; i <= endHi; ++i) {
             if (ranges[i].matches(value)) {
-                        collectBucket(sub, doc, subBucketOrdinal(owningBucketOrdinal, i));
+                collectBucket(sub, doc, subBucketOrdinal(owningBucketOrdinal, i));
             }
         }
 
         return endHi + 1;
     }
+
+    /**
+     * Attempt to collect these ranges via the BKD tree instead of DocValues.
+     *
+     * This estimates the number of matching points in the BKD tree.  If it is
+     * less than 75% of maxDoc we attempt to use the BKD tree to collect values.
+     * The BKD tree is potentially much faster than DV collection because
+     * we only need to inspect leaves that overlap each range, rather than
+     * collecting all the values as with DVs.  And since we only care about doc
+     * counts, we don't need to decode values when an entire leaf matches.
+     *
+     * If we use the BKD tree, when it is done collecting values a
+     * {@link CollectionTerminatedException} is thrown to signal completion
+     */
+    private void tryBKDOptimization(LeafReaderContext ctx, LeafBucketCollector sub) throws CollectionTerminatedException, IOException {
+        final PointValues pointValues = ctx.reader().getPointValues(pointField);
+        if (pointValues != null) {
+            PointValues.IntersectVisitor[] visitors = new PointValues.IntersectVisitor[ranges.length];
+            DocIdSetBuilder[] results = new DocIdSetBuilder[ranges.length];
+
+            final Bits liveDocs = ctx.reader().getLiveDocs();
+            int maxDoc = ctx.reader().maxDoc();
+            long estimatedPoints = 0;
+            for (int i = 0; i < ranges.length; i++) {
+                // OK to allocate DocIdSetBuilder now since it allocates memory lazily and the
+                // estimation won't call `grow()` on the visitor (only once we start intersecting)
+                results[i]  = new DocIdSetBuilder(maxDoc);
+                visitors[i] = getVisitor(liveDocs, encodedRanges[i * 2], encodedRanges[i * 2 + 1], results[i]);
+                estimatedPoints += pointValues.estimatePointCount(visitors[i]);
+            }
+
+            if (estimatedPoints < maxDoc * 0.75) {
```

Comment:
It's not clear to me what we should actually set this threshold to.  75% might be too aggressive, since the BKD estimate could under-estimate some leaves considerably.

Reply:
I wonder if `maxDoc` is the right parameter as we are counting values. Maybe we should be using `pointValues.size()` instead?

Is this heuristic extracted from performance tests?

Reply:
Nothing very scientific at all, happy to change it for something better :)

It was derived from performance tests only in the sense that ranges which span most of the data tend to be same speed (or slower) and with a bit more overhead than just collecting them via DocValues.

The rationale was that the DocValue approach would have to touch `maxDocs` documents in the collect loop, and so I wanted to "collect" fewer than that many values from the BKD tree.  But it's not very comparable since they have different costs (and even in the tree, collecting all the docs from one leaf is much different than when we have to unpack individual values)